maven/lib/dependabot/maven/version_parser.rb

```diff
@@ -0,0 +1,224 @@
+# typed: strict
+# frozen_string_literal: true
+
+require "sorbet-runtime"
+require "strscan"
+
+# See https://maven.apache.org/pom.html#Version_Order_Specification for details
+#
+module Dependabot
+  module Maven
+    class TokenBucket < T::Struct
+      extend T::Sig
+      extend T::Helpers
+      include Comparable
+
+      prop :tokens, T::Array[T.untyped]
+      prop :addition, T.nilable(TokenBucket)
+
+      sig { returns(T::Array[T.untyped]) }
+      def to_a
+        return tokens if addition.nil?
+
+        tokens.clone.append(addition.to_a)
+      end
+
+      sig { params(other: TokenBucket).returns(T.nilable(Integer)) }
+      def <=>(other)
+        cmp = compare_tokens(tokens, other.tokens)
+        return cmp unless cmp&.zero?
+
+        compare_additions(addition, other.addition)
+      end
+
+      sig do
+        params(
+          first: T::Array[T.any(String, Integer)],
+          second: T::Array[T.any(String, Integer)]
+        ).returns(T.nilable(Integer))
+      end
+      def compare_tokens(first, second)
+        max_idx = [first.size, second.size].max - 1
+        (0..max_idx).each do |idx|
+          cmp = compare_token_pair(first[idx], second[idx])
+          return cmp unless T.must(cmp).zero?
+        end
+        0
+      end
+
+      sig do
+        params(
+          first: T.nilable(T.any(String, Integer)),
+          second: T.nilable(T.any(String, Integer))
+        ).returns(T.nilable(Integer))
+      end
+      def compare_token_pair(first = 0, second = 0) # rubocop:disable Metrics/PerceivedComplexity,Metrics/CyclomaticComplexity
+        first ||= 0
+        second ||= 0
+
+        if first.is_a?(Integer) && second.is_a?(String)
+          return first <= 0 ? -1 : 1
+        end
+
+        if first.is_a?(String) && second.is_a?(Integer)
+          return second <= 0 ? 1 : -1
+        end
+
+        if first == Dependabot::Maven::VersionParser::SP &&
+           second.is_a?(String) && second != Dependabot::Maven::VersionParser::SP
+          return -1
+        end
+
+        if second == Dependabot::Maven::VersionParser::SP &&
+           first.is_a?(String) && first != Dependabot::Maven::VersionParser::SP
+          return 1
+        end
+
+        if first.is_a?(Integer) && second.is_a?(Integer)
+          first <=> second
+        elsif first.is_a?(String) && second.is_a?(String)
+          first <=> second
+        end
+      end
+
+      sig do
+        params(first: T.nilable(TokenBucket), second: T.nilable(TokenBucket)).returns(T.nilable(Integer))
+      end
+      def compare_additions(first, second)
+        return 0 if first.nil? && second.nil?
+
+        (first || empty_addition) <=> (second || empty_addition)
+      end
+
+      sig { returns(TokenBucket) }
+      def empty_addition
+        TokenBucket.new(tokens: [])
+      end
+    end
+
+    class VersionParser
```

Comment:
style question - any reason to have multiple classes in a single file? In other languages, the convention was to have one class per file but I don't know enough Ruby.

tagging @landongrindheim to weigh in.

Reply:
It's 1 of those cases where I don't have strong preference either way. Initially, I just wanted a ruby struct to hold the data for me. Sorbet had other ideas. It recommended subclassing T::Struct or having a standalone class. But these classes are tightly coupled that's why I was a bit hesitant to put them in separate files. If I had known earlier that sorbet would make me do it this way I'd probably have gone with 1 class per file.

Reply:
IMO, it would be more conventional to define `TokenBucket` in a `token_bucket.rb`, rather than in the same file as another object of the same nesting.

If `TokenBucket` is specific to `Maven::VersionParser`, nesting it inside `Maven::VersionParser` is reasonable. The footprint of the object is small, but the LOC (mostly due to its typings) is large enough that I would probably move it to another file. Your choice 😄

Reply:
Thanks @landongrindheim I'll move it to a separate file.